src/svd/path.py

```diff
@@ -0,0 +1,298 @@
+#
+# Copyright (c) 2023 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+"""
+Classes for refererencing SVD elements based on name.
+"""
+
+from __future__ import annotations
+
+import re
+from abc import ABC, abstractmethod
+from functools import singledispatch
+from itertools import chain
+from typing import (
+    Any,
+    Generic,
+    Iterable,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+    TypeVar,
+    Union,
+    overload,
+)
+from typing_extensions import Self
+
+
+PartT = TypeVar("PartT")
+
+
+class AbstractSPath(ABC, Sequence[PartT]):
+    """
+    A type used to describe paths of named SVD elements.
+    """
+
+    @abstractmethod
+    def __init__(self, *parts: Union[PartT, Sequence[PartT]]) -> None:
+        """
+        :param parts: Path segments
+        """
+        ...
+
+    @property
+    @abstractmethod
+    def parts(self) -> Tuple[PartT, ...]:
+        """:return: Path components."""
+        ...
+
+    @property
+    @abstractmethod
+    def name(self) -> Optional[str]:
+        """:return: Name of the element pointed to by the path, including any array indices."""
+        ...
+
+    @property
+    @abstractmethod
+    def stem(self) -> Optional[str]:
+        """:return: Name of the element pointed to by the path, excluding any array indices"""
+        ...
+
+    @property
+    @abstractmethod
+    def parent(self) -> Optional[AbstractSPath]:
+        """:return: Path to the parent element of this path, if it exists."""
+        ...
+
+    def join(self, *other: Union[PartT, Sequence[PartT]]) -> Self:
+        """:return: The path resulting from appending other to the end of this path. """
+        return self.__class__(*self.parts, *other)
+
+    @overload
+    def __getitem__(self, item: int, /) -> PartT:
+        ...
+
+    @overload
+    def __getitem__(self, item: slice, /) -> Self:
+        ...
+
+    def __getitem__(self, item: Union[int, slice], /) -> Union[PartT, Self]:
+        if isinstance(item, slice):
+            return self.__class__(*self.parts[item])
+        else:
+            return self.parts[item]
+
+    def __len__(self) -> int:
+        return len(self.parts)
+
+    def __hash__(self) -> int:
+        return hash(self.parts)
+
+    def __eq__(self, other: Any) -> bool:
+        return self.parts == other
+
+    @abstractmethod
+    def __repr__(self) -> str:
+        ...
+
+
+class FSPath(AbstractSPath[str]):
```

Comment:
nit: What about `EPath` and `FEPath` instead for "element and "flat element" respectively? Given that `svada` is for parsing SVDs, that adjective is a given; and using `S` for `SVD` seems inaccurate when these are element paths, not SVD paths.